Automattic/jetpack
# auto activate userless modules

As part of the User-less project, we need to activate default modules when the site connects at a site level and should not wait until a user connects, as we are doing now.

This PR implements the following behavior:
* When the blog first connects, it activates all the default modules that do not require a user connection
* When a user is first connected, it activates all the remaining default modules
* On reconnection, re-activates all the modules that were active before disconnection.

#### Changes proposed in this Pull Request:

* add filters to `get_available_modules` and `get_default_modules` to be able to get only modules that require or not require a user connection
* Activate default modules that do no require user connection when the site is first connected at a site level
* Activate the remaining default modules when a user is connected for the first time
* Make sure that on reconnection whe re-activate the modules that were active before disconnection (current behavior)

#### A note on automated tests

All this part of the code could use a big refactor for us to be able to better test it. I started doing it but the rabbit hole started to grow and I dropped it, because a simple refactor (moving the methods to a new dedicated class) would not be of much value.

So I wrote a few tests just to make sure the added filters work, but they are limited tests until we seriously refactor it.

#### Jetpack product discussion


Link soon

#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






* Start with a fresh connection
* Set the `JETPACK_NO_USER_TEST_MODE` constant to `true`
* Use this command to check the list of active modules `wp jetpack options get active_modules`
* Make sure this option does not exist yet

* Connect the site, but don't connect a user
* Check active modules and make sure you get
```
(
    [0] => contact-form
    [1] => enhanced-distribution
    [2] => json-api
    [3] => protect
    [4] => stats
    [5] => verification-tools
    [6] => woocommerce-analytics
)
```

* Connect the user and check the active modules again
* Check that `notes` was added to the list
```
    ...
    [7] => notes
```

* On the UI, deactivate "Protect" and activate "Monitor"
* check the list of active modules and make sure it reflects the changes

* Disconnect and connect the site again, only at a site level
* Check the active modules and make sure the list is still the same it was before disconnecting
* Protect should not be re-enabled and Monitor must be enabled
* Check the UI correctly represents the current state (module active, but ask for user connection)

![Captura de Tela 2021-02-24 às 11 20 40](https://user-images.githubusercontent.com/971483/109053907-01240880-76bc-11eb-8b78-250e8c0238a2.png)


* Connect a user and make sure nothing changes

* Play around with different combinations of activating/deactivating modules, connecting and disconnecting, and make sure it behaves in the way described at the top of this PR

Now a very specific test:

* Disconnect your site
* Reset options: `wp jetpack options delete active_modules` and `wp jetpack options delete active_modules_initialized`
* Connect at a site level
* Deactivate Protect
* Connect the user
* Make sure `notes` was activated but `protect` was not re-activated


#### Proposed changelog entry for your changes:


* Activate default modules when the site connects on a site level
